Add local version extension

drewr opened the pull request:
In cases where we need to publish a new version of our JNA jar apart from upstream, we need a way to identify it from the last.  This implements a scheme similar to the way OS packages do it, where we add an ordinal after the major.minor.patch ("pre-release" in semver).

drewr committed:
Add local version extension to support downstream changes

We need to bump our local version to enable aarch64, but upstream
net.java.dev.jna has not moved. Introduce a version
extension ("pre-release" in semver) to increment.

nik9000 commented:
I tried to lookup what debian calls these, and, unhelpfully, they call it the ["debian_revision"](https://www.debian.org/doc/debian-policy/ch-controlfields.html#s-f-Version).

rjernst commented:
Can we only do this when there is a change actually needed? ie we only add the `-1` suffix when we need to patch a released version.

drewr commented:
@rjernst Are you proposing just manually patch the code when this needs to be done as opposed to changing the versioning scheme more permanently?

rjernst commented:
I'm saying the first release for a given "real" jna version would match the jna version. Then, if in this rare scenario we need to modify our jar (eg adding more architectures like in this case), we append "-1", "-2" and so on to the version (that we release). The version in jna would always stay the same, this would just be in our maven artifacts.

drewr committed:
Add project url for pom, Maven Central requires it

drewr committed:
Use an optional `prerelease` flag instead of hardcoding

To enable the ability to upload to maven a different version from the
upstream net.java.dev/jna.

drewr committed:
Add instructions for releasing a modified version to Maven

drewr commented:
@rjernst I've added a `prerelease` flag (naming suggestions welcome) that will optionally enable this behavior if needing to publish an alternate to maven central only. The full story is in the [readme](https://github.com/elastic/jna-build/blob/2ea060dd5929f53370d14a428c6f773e5e73299e/README.txt#L4-L49).

drewr committed:
Use variable for upstream version as well

drewr commented:
Actually, [this version of the readme](https://github.com/elastic/jna-build/blob/4ea6d0003211efdbb756d1526aeacc80acfd3587/README.txt#L4-L50) makes more sense.

rjernst commented:
> I've added a prerelease flag (naming suggestions welcome)

I don't think this has anything to do with prerelease (which to me implies alpha/beta/preview type naming). It's more a "version suffix"?

And thank you for the readme! I think it only works, though, when a suffix is present? I would just have the suffix include the dash.

rjernst commented on build.gradle:
```diff
@@ -136,6 +144,7 @@ publishing {
         node.appendNode('name', 'Elastic JNA Distribution')
         node.appendNode('description',
           'A build of jna which supports all platforms supported by Elasticsearch')
+        node.appendNode('url', 'https://github.com/elastic/jna-build')
```
Why is this necessary? I don't think we should use our own repo because we are just repackaging jna. But I was also able to publish this to maven central before without adding it (at least I don't remember modifying the pom manually...).

drewr replied:
It kept failing without it when I tried to close the repository.  I can regenerate and take a screenshot.

I thought ours would be more appropriate since we actually packaged it, I'm fine with the upstream. Will change.

drewr commented:
I'll rework the readme for the more general case and switch to `SUFFIX` for the name.

drewr committed:
Update publishing docs for the normal case

drewr committed:
Use markdown instead

drewr committed:
Use upstream URL

drewr committed:
Title

drewr commented:
It's a little verbose, but I updated the docs to cover the default case first with a bit more explanation.

rjernst commented on README.md:
```diff
@@ -0,0 +1,112 @@
+Then build:
+
+    % vagrant up
+    % gradle assemble generatePomFileForJnaPublication
```
Just `gradle assemble` should be sufficient here (assemble is set to depend on the pom generation task).

drewr replied:
Great, didn't realize the dep there.

rjernst commented on README.md:
```diff
@@ -0,0 +1,112 @@
+javadoc jars as well.  We don't have any automation here to build
+them, so simply reuse from net.java.dev/jna:
+
+    % mkdir -p tmp
```
I would put this under `build/tmp`, instead of creating another dir that must exist in `.gitignore`.

drewr replied:
Even better, just used `build`.

rjernst commented on build.gradle:
```diff
@@ -2,7 +2,19 @@
+// the local version.  See readme for details.
+String versionSuffix = System.getProperty("version.suffix")
+
+if (versionSuffix != null) {
```
You can avoid this conditional by using a default value of "" in the getProperty call above.

drewr replied:
Of course, didn't realize that version. Thanks.

rjernst reviewed (approved):
LGTM. A few last nits.

drewr committed:
Remove unneeded conditional

drewr committed:
Update docs

drewr commented:
Thanks @rjernst!
